src/sinks/kafka.rs

```diff
@@ -338,13 +348,17 @@ impl Sink<Event> for KafkaSink {
         let this = Pin::into_inner(self);
         while !this.in_flight.is_empty() {
             match ready!(Pin::new(&mut this.in_flight).poll_next(cx)) {
-                Some((seqno, Ok(result))) => {
+                Some((seqno, Ok(result), metadata)) => {
                     match result {
                         Ok((partition, offset)) => {
-                            trace!(message = "Produced message.", ?partition, ?offset)
+                            metadata.update_status(EventStatus::Delivered);
+                            trace!(message = "Produced message.", ?partition, ?offset);
+                        }
+                        Err(error) => {
+                            metadata.update_status(EventStatus::Errored); // should this be `Failed`?
```

Comment:
If `Errored` implies a problem with the data and `Failed` implies a problem with the server, this should probably be `Failed`. Kafka messages are just bytes so it's hard to make one invalid.

Reply:
It's more an issue of temporary vs permanent errors. That is, `Errored` indicates the events may be retried but `Failed` indicates permanent rejection.

Reply:
In that case I'd probably lean towards `Errored`. I'm not entirely sure what a "permanent rejection" would entail for Kafka (maybe a message that's too large?).